src/providers/postgres/qgspostgreslistener.cpp

```diff
@@ -58,7 +61,43 @@ QgsPostgresListener::~QgsPostgresListener()
 void QgsPostgresListener::run()
 {
   PGconn *conn = nullptr;
-  conn = PQconnectdb( mConnString.toLocal8Bit() );
+  QString connectString = mConnString;
+
+  connectString += QStringLiteral( " connect_timeout=%1" ).arg( PG_CONNECT_TIMEOUT );
+  conn = PQconnectdb( connectString.toUtf8() );
+
+  if ( PQstatus( conn ) != CONNECTION_OK )
+  {
+    QgsDataSourceUri uri( connectString );
+    QString username = uri.username();
+    QString password = uri.password();
+
+    PQfinish( conn );
+
+    QgsCredentials::instance()->lock();
+
+    if ( QgsCredentials::instance()->get( mConnString, username, password, PQerrorMessage( conn ) ) )
+    {
+      uri.setUsername( username );
+      uri.setPassword( password );
+      connectString = uri.connectionInfo( false );
+      connectString += QStringLiteral( " connect_timeout=%1" ).arg( PG_CONNECT_TIMEOUT );
+
+      conn = PQconnectdb( connectString.toUtf8() );
+      if ( PQstatus( conn ) == CONNECTION_OK )
+        QgsCredentials::instance()->put( mConnString, username, password );
```

Comment:
Why are you putting the credentials back to the credentials manager they are from?

Reply:
If I understand correctly the main idea of QgisCredentials: Get -> Use -> (if everything is OK, returned for further use) Put.

Reply:
Ok just because the `get` may not have the credentials and ask them to the user, you have to save them in the credentials manager. Got it!